diff -up geoclue-0.12.0/src/master-provider.c.blah geoclue-0.12.0/src/master-provider.c
--- geoclue-0.12.0/src/master-provider.c.blah	2011-03-07 17:32:55.926713653 -0600
+++ geoclue-0.12.0/src/master-provider.c	2011-03-07 17:33:02.177635505 -0600
@@ -846,10 +846,6 @@ gc_master_provider_initialize_interfaces
 static gboolean
 gc_master_provider_initialize (GcMasterProvider *provider)
 {
-	GcMasterProviderPrivate *priv;
-	 
-	priv = GET_PRIVATE (provider);
-	
 	if (!gc_master_provider_initialize_interfaces (provider)) {
 		return FALSE;
 	}
